### PR TITLE
Changing  MPC_DATA_CONTAINER type to a boost shared pointer 

### PR DESCRIPTION
**Pull request by adityaghantasala**
Major change is to change the type of MPC_DATA_CONTAINER to a boost shared pointer to avoid memory leaks and seg faults.

@philbucher  can you check if this commit is still causing seg faults you mentioned.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h [hunk 1]
```diff
@@ -88,13 +89,13 @@ class ApplyMultipointConstraintsProcess : public Process
         // IMPORTANT : This constructor is not to be used when using this process in the normal KRATOS process_list of python script
         ProcessInfoPointerType info = mr_model_part.pGetProcessInfo();
         if (info->GetValue(MPC_DATA_CONTAINER) == NULL)
```
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h]: This shouldn't be a `nullptr`?

**Review by pooyan-dadvand — COMMENTED** [review]
Seems ok to me! :+1: 

Except the minor null check. you can also use !mpcPointer to check if is null or not.

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h (hunk 1 above)
Comment (adityaghantasala) [reply applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h]: Thank you for the hint Pooyan. I changed it and pushed the commit.

**Comment by philbucher** [comment]
I get a linker error with the latest commit:
```
CMakeFiles/KratosStructuralMechanicsApplication.dir/structural_mechanics_application.cpp.o: In function `Kratos::KratosStructuralMechanicsApplication::Register()':
/home/philippb/software/Kratos/applications/StructuralMechanicsApplication/structural_mechanics_application.cpp:202: undefined reference to `Kratos::LOCAL_INERTIA_VECTOR'
CMakeFiles/KratosStructuralMechanicsApplication.dir/custom_python/structural_mechanics_python_application.cpp.o: In function `Kratos::Python::init_module_KratosStructuralMechanicsApplication()':
/home/philippb/software/Kratos/applications/StructuralMechanicsApplication/custom_python/structural_mechanics_python_application.cpp:79: undefined reference to `Kratos::LOCAL_INERTIA_VECTOR'
CMakeFiles/KratosStructuralMechanicsApplication.dir/custom_elements/cr_beam_element_3D2N.cpp.o: In function `Kratos::Variable<boost::numeric::ublas::vector<double, boost::numeric::ublas::unbounded_array<double, std::allocator<double> > > >::Type& Kratos::Properties::GetValue<Kratos::Variable<boost::numeric::ublas::vector<double, boost::numeric::ublas::unbounded_array<double, std::allocator<double> > > > >(Kratos::Variable<boost::numeric::ublas::vector<double, boost::numeric::ublas::unbounded_array<double, std::allocator<double> > > > const&)':
/home/philippb/software/Kratos/kratos/includes/properties.h:182: undefined reference to `Kratos::LOCAL_INERTIA_VECTOR'
/home/philippb/software/Kratos/kratos/includes/properties.h:182: undefined reference to `Kratos::LOCAL_INERTIA_VECTOR'
/home/philippb/software/Kratos/kratos/includes/properties.h:182: undefined reference to `Kratos::LOCAL_INERTIA_VECTOR'
CMakeFiles/KratosStructuralMechanicsApplication.dir/custom_elements/cr_beam_element_3D2N.cpp.o:/home/philippb/software/Kratos/kratos/includes/properties.h:282: more undefined references to `Kratos::LOCAL_INERTIA_VECTOR' follow
collect2: error: ld returned 1 exit status
applications/StructuralMechanicsApplication/CMakeFiles/KratosStructuralMechanicsApplication.dir/build.make:1191: recipe for target 'applications/StructuralMechanicsApplication/KratosStructuralMechanicsApplication.so' failed
make[2]: *** [applications/StructuralMechanicsApplication/KratosStructuralMechanicsApplication.so] Error 1
CMakeFiles/Makefile2:1022: recipe for target 'applications/StructuralMechanicsApplication/CMakeFiles/KratosStructuralMechanicsApplication.dir/all' failed
make[1]: *** [applications/StructuralMechanicsApplication/CMakeFiles/KratosStructuralMechanicsApplication.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by adityaghantasala** [comment]
@philbucher  I added this missing variable. Can you check now ?

**Comment by philbucher** [comment]
It compiles but the test still fails sometimes, sometimes it works.
If it fails this is the error:
```
======================================================================
FAIL: test_MPC_Constraints (__main__.TestMultipointConstraintsTwo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_multipoint_contstraints.py", line 237, in test_MPC_Constraints
    self._check_results(mp)
  File "test_multipoint_contstraints.py", line 245, in _check_results
    self.assertAlmostEqual(disp1, disp2, 5)
AssertionError: 0.0 != nan within 5 places

----------------------------------------------------------------------
```
The assertion error can also be: `AssertionError: nan != nan within 5 places`

**Comment by philbucher** [comment]
btw I recommend using the `VariableUtils` from the core to add and fix dofs and set a variable to 0:
* `SetToZero_ScalarVar`
* `ApplyFixity`
* `AddDof`

I mean in your tests

**Comment by adityaghantasala** [comment]
@philbucher  can you check it now ?

**Comment by philbucher** [comment]
Now it works for me, but you might want to remove the prints

File: applications/StructuralMechanicsApplication/tests/test_multipoint_contstraints.py [hunk 2]
```diff
@@ -108,230 +123,171 @@ def _solve(self):
...
-        while(time <= end_time):
+        while (time <= end_time):
             time = time + dt
             step = step + 1
-            print('############ Time :: ', time, ' ### step ', step)
             mp.CloneTimeStep(time)
-
-            self._solve()
+            if (step > 3):
+                print('############################### Time :: ', time,
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_multipoint_contstraints.py]: Protect that by an echo lv?
I think one can set some verbosity in the tests

**Review by philbucher — APPROVED** [review]

File: applications/StructuralMechanicsApplication/tests/test_multipoint_contstraints.py (hunk 2 above)
Comment (adityaghantasala) [reply applications/StructuralMechanicsApplication/tests/test_multipoint_contstraints.py]: I will remove it I think as even with a echo lvl I dont see it is necessary.!

Comment (adityaghantasala) [reply applications/StructuralMechanicsApplication/tests/test_multipoint_contstraints.py]: Thanks !

**Pull request merged**